File: app/src/main/java/com/damianmichalak/shopping_list/model/HistoryDao.java
package com.damianmichalak.shopping_list.model;

import com.damianmichalak.shopping_list.helper.Database;
import com.damianmichalak.shopping_list.helper.References;
import com.damianmichalak.shopping_list.model.api_models.Product;
import com.google.firebase.database.DatabaseReference;

import java.util.Map;

import javax.annotation.Nonnull;
import javax.inject.Inject;
import javax.inject.Singleton;

import rx.Observable;

@Singleton
public class HistoryDao {

    @Nonnull
    private final Database<Product> database;
    @Nonnull
    private final Observable<DatabaseReference> referenceObservable;
    @Nonnull
    private final Observable<Map<String, Product>> productsObservable;

    @Inject
    public HistoryDao(@Nonnull final References references,
                      @Nonnull final Database<Product> database,
                      @Nonnull final CurrentListDao currentListDao) {
        this.database = database;

        referenceObservable = currentListDao.getCurrentListKeyObservable()
                .filter(uid -> uid != null)
                .map(references::historyReference)
                .replay(1)
                .refCount();

        productsObservable = referenceObservable.switchMap(reference -> database.itemsAsMap(reference, Product.class, "datePurchased"))
                .replay(1)
                .refCount();
    }

    Observable<Boolean> addNewItemObservable(final Product product) {
        return referenceObservable.switchMap(reference -> database.put(product.purchased(), reference));
    }

    public Observable<Boolean> removeItemByKeyObservable(final String key) {
        return referenceObservable.switchMap(reference -> database.remove(key, reference));
    }

    @Nonnull
    public Observable<Map<String, Product>> getProductsObservable() {
        return productsObservable;
    }
}

File: app/src/main/java/com/damianmichalak/shopping_list/model/HistoryDao.kt
package com.damianmichalak.shopping_list.model

import com.damianmichalak.shopping_list.helper.Database
import com.damianmichalak.shopping_list.helper.References
import com.damianmichalak.shopping_list.model.apiModels.Product
import com.google.firebase.database.DatabaseReference
import javax.inject.Inject
import javax.inject.Singleton

import rx.Observable

@Singleton
class HistoryDao @Inject
constructor(references: References,
            private val database: Database<Product>,
            currentListDao: CurrentListDao) {
    private val referenceObservable: Observable<DatabaseReference>
    val productsObservable: Observable<Map<String, Product>>

    init {

        referenceObservable = currentListDao.currentListKeyObservable
                .filter { uid -> uid != null }
                .map { references.historyReference(it) }
                .replay(1)
                .refCount()

        productsObservable = referenceObservable.switchMap { reference -> database.itemsAsMap(reference, Product::class.java, "datePurchased") }
                .replay(1)
                .refCount()
    }

    fun addNewItemObservable(product: Product): Observable<Boolean> {
        return referenceObservable.switchMap { reference -> database.put(product.purchased(), reference) }
    }

    fun removeItemByKeyObservable(key: String): Observable<Boolean> {
        return referenceObservable.switchMap { reference -> database.remove(key, reference) }
    }
}
